### DISCUSSION TITLE
Single-file scripts autoinstall dependencies support

### DISCUSSION THREAD

**Discussion by jcrben** [nodejs/help#4635] [Q&A] [answered]
Bun and deno allow for writing imports in single-file scripts without a package.json - in bun I believe this is called auto-install or autoimport  with the Bun module resolution algorithm https://bun.sh/docs/runtime/autoimport

This aligns with similar features in various languages https://dbohdan.com/scripts-with-dependencies#javascript - `@Grab` from Groovy is a relatively old flavor of this feature, with Python's `pip-run` being more recent

I tried to find an issue proposing this feature but I couldn't - has this been discussed? I figured searching bun or deno would be mentioned in such an issue and I didn't find them.

**Comment by avivkeller** [comment]
CC relevant teams: @nodejs/loaders @nodejs/npm

**Comment by mcollina** [comment, marked as answer]
I'm sorry, but the explanation of "why not" is likely pretty long, and I have no real time for it. It boils down to security and reproducibility, so I don't expect this feature to be implemented anytime soon.

**Reply by avivkeller** [reply]
Additionally, it requires Node.js/npm to have an on-the-fly caching system, which isn't an easy feat to implement...

**Reply by jcrben** [reply]
Appreciate the response and understood. In terms of reproducibility, `bun` allows for specifying a version: https://bun.sh/docs/runtime/autoimport#version-specifier (other languages work similarly such as `pip-run` and `@Grab`). For security, people can point to a private registry and manage what is in that private registry. But granted not sure how much this might mess with standards in the language.
